packages/devtools_app/lib/src/timeline/timeline_controller.dart

```diff
@@ -116,70 +124,114 @@ class TimelineController {
     _cpuProfileTransformer.processData(cpuProfileData);
   }
 
-  void loadOfflineData(OfflineTimelineData offlineData) {
-    // TODO(kenz): loading offline data should respect user's current timeline
-    // mode (frameBased vs full). It should also support toggling modes.
-    frameBasedTimeline.data = offlineData.copy();
-    offlineTimelineData = offlineData.copy();
-
-    final traceEvents =
-        offlineData.traceEvents.map((trace) => TraceEvent(trace)).toList();
+  void loadOfflineData(TimelineData offlineData) {
+    final traceEvents = offlineData.traceEvents
+        .map((trace) => TraceEventWrapper(
+              TraceEvent(trace),
+              DateTime.now().microsecondsSinceEpoch,
+            ))
+        .toList();
 
     // TODO(kenz): once each trace event has a ui/gpu distinction bit added to
-    // the trace, we will not need to infer thread ids. Since we control the
-    // format of the input, this is okay for now.
-    final uiThreadId = traceEvents.first.threadId;
-    final gpuThreadId = traceEvents.last.threadId;
-
-    frameBasedTimeline.processor = FrameBasedTimelineProcessor(
-      timelineController: this,
-      uiThreadId: uiThreadId,
-      gpuThreadId: gpuThreadId,
-    );
-
-    for (TraceEvent event in traceEvents) {
-      frameBasedTimeline.processor.processTraceEvent(
-        TraceEventWrapper(event, DateTime.now().millisecondsSinceEpoch),
-        immediate: true,
+    // the trace, we will not need to infer thread ids. This is not robust.
+    final uiThreadId = traceEvents
+            .firstWhere((trace) => trace.event.name == uiEventName,
+                orElse: () => null)
+            ?.event
+            ?.threadId ??
+        -1;
```

Comment:
add a constant that is invalidThreadId or something. `-1` seems a bit arbitrary.